src/database.ts

```diff
@@ -160,214 +149,94 @@ export class Database {
-    sql: string,
-    args: Record<string, BindValue>,
-  ): T[];
-  queryArray<T extends unknown[] = any[]>(
-    sql: string | TemplateStringsArray,
-    ...args: BindValue[] | [Record<string, BindValue>]
-  ): T[] {
-    const stmt = this.prepare(typeof sql === "string" ? sql : sql.join("?"));
-    if (isObject(args[0])) {
-      stmt.bindAllNamed(args[0] as Record<string, BindValue>);
-    } else {
-      stmt.bindAll(...args as BindValue[]);
-    }
-    const rows = [];
-    for (const row of stmt) {
-      rows.push(row.asArray());
-    }
-    stmt.finalize();
-    return rows as T[];
+  run(sql: string, ...params: RestBindParameters): void {
+    this.exec(sql, ...params);
   }
 
-  /**
-   * Executes an SQL query and returns the rows as objects.
-   *
-   * Note: if you do not need the column names, consider calling `queryArray` instead.
-   * As this method does an extra FFI call to get the column names, it is more expensive than `queryArray`.
-   *
-   * Example:
-   * ```ts
-   * const users = db.queryObject<{
-   *   id: number,
-   *   username: string,
-   * }>("select id, username from users");
-   *
-   * // Using bind parameters
-   * const [user] = db.queryObject<{
-   *   id: number,
-   *   username: string,
-   * }>("select id, username from users where email = ?", email);
-   *
-   * // Using template strings
-   * const [user] = db.queryObject<{
-   *  id: number,
-   *  username: string,
-   * }>`select id, username from users where email = ${email}`;
-   *
-   * // Using named bind parameters
-   * const [user] = db.queryObject<{
-   *   id: number,
-   *   username: string,
-   * }>("select id, username from users where email = :email", { email });
-   * ```
-   *
-   * @param sql SQL query to execute.
-   * @param args Parameters to bind to the query.
-   *
-   * @returns Array of rows, where rows are objects mapping column names to values.
-   */
-  queryObject<T extends Record<string, unknown> = Record<string, any>>(
-    strings: TemplateStringsArray,
-    ...args: BindValue[]
-  ): T[];
-  queryObject<T extends Record<string, unknown> = Record<string, any>>(
-    sql: string,
-    ...args: BindValue[]
-  ): T[];
-  queryObject<T extends Record<string, unknown> = Record<string, any>>(
-    sql: string,
-    args: Record<string, BindValue>,
-  ): T[];
-  queryObject<T extends Record<string, unknown> = Record<string, any>>(
-    sql: string | TemplateStringsArray,
-    ...args: BindValue[] | [Record<string, BindValue>]
-  ): T[] {
-    const stmt = this.prepare(typeof sql === "string" ? sql : sql.join("?"));
-    if (isObject(args[0])) {
-      stmt.bindAllNamed(args[0] as Record<string, BindValue>);
-    } else {
-      stmt.bindAll(...args as BindValue[]);
-    }
-    const rows = [];
-    for (const row of stmt) {
-      rows.push(row.asObject());
+  #cachedQueriesKeys: string[] = [];
+  #cachedQueriesLengths: number[] = [];
+  #cachedQueriesValues: Statement[] = [];
+
+  query(sql: string): Statement {
```

Comment:
I'm not quite sure about this method yet. The caching seems bit weird and the case where user manually finalizes the cached statement yet it will be used again. Maybe we can do this in follow ups?

Reply:
yeah in favour of doing in a follow up